### PR TITLE
 Fix EZP-25003: clear HTTP cache when trashing and recovering content

### PR DESCRIPTION
**Pull request by bdunogier**
> Fixes http://jira.ez.no/browse/EZP-25003

~~Clears all HTTP cache (like move subtree does) when a location is trashed. Maybe it could be restricted, but this seems to fall in the same case than delete, where the signal is sent _after_ the item has been deleted.~~~

Done:
- Adds TrashSlot and RecoverSlot
- Adds affected parentLocation on Trash and DeleteLocation signal, pass this to purge function
- Adds affected contentId on Recover signal so cache clearing can clear relevant cache.

Also done but can be moved to other PR if needed:
- Avoid exception on DeleteContent signal *(for now, afaik UI exposes Trash so should not be to common)* by purging all cache as we don't have affected locations atm, solution is multi tag and just clear content id related tags.
- Some cleanup on argument name for purge().

### TODO
- [ ] Confirm fix by manual testing
- [ ] REST behat coverage
- [ ] unit tests coverage for new slots

### CODE REVIEW

File: eZ/Publish/Core/MVC/Symfony/Cache/Http/SignalSlot/TrashSlot.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * @version //autogentag//
+ */
+namespace eZ\Publish\Core\MVC\Symfony\Cache\Http\SignalSlot;
+
+use eZ\Publish\Core\SignalSlot\Signal;
+
+/**
+ * A slot handling TrashSignal.
+ */
+class TrashSlot extends PurgeAllHttpCacheSlot
```
Comment (andrerom) [inline eZ/Publish/Core/MVC/Symfony/Cache/Http/SignalSlot/TrashSlot.php]: Is the naming off or does it actually purge all http cache and basicly DDOS the site? :)

Comment (andrerom) [inline eZ/Publish/Core/MVC/Symfony/Cache/Http/SignalSlot/TrashSlot.php]:
```php
    public function purgeAll()
    {
        $this->cacheManager->invalidate(array('X-Location-Id' => '.*'));
    }
```

splendid..

**Comment by andrerom** [comment]
I know it fixes the qa bug, but I'm tempting to veto this, we obviously need to start to tackle all the technical debt left behind here in a proper way.

This is just QA complaining for now right? So we can ask them to lower cache_ttl and that should kind of workaround it.

File: eZ/Publish/Core/MVC/Symfony/Cache/Http/SignalSlot/TrashSlot.php (hunk 1 above)
Comment (bdunogier) [inline eZ/Publish/Core/MVC/Symfony/Cache/Http/SignalSlot/TrashSlot.php]: It is precisely what the MoveSlot does, for information.

**Comment by bdunogier** [comment]
Thank you for the feedback.

**Comment by andrerom** [comment]
Summary of discussion:
- keep the new Slot you added
- however also add a new `PurgeForLocationHttpCacheSlot`and use it instead here and in `MoveSubtreeSlot` and `UnnassignUserFromUserGroupSlot`, deprecate `PurgeAllHttpCacheSlot`. You'll end up using GatewayCachePurger->purge() for this which is deprecated, reason why it is deprecated is that it can not take advantage of the Cache clear rules using the `ContentCacheClearEvent` stuff.

Followup story to solve it fully, options:
- Either we change the API on such operations to be a command *(start move towards CQRS)*, and lets events/slots enhance the command object so it can add meta info about affected location id's for instance which is used after the operation itself to clear cache.
-- Alternatively as part of this we handle commands async, so then we can handle operation in batch and get it to emit events for every single location deleted, meaning bug fix done now will be useful as we discussed above.
- Or we solve this with multi tagging, however this is progressing slowly at FOS.
- Or we can as a bit smaller effort add array for affected location id's on signals affecting trees, and get SignalSlot repo to call a new spi/api to get location id's in a tree before the itself operation.

**Comment by andrerom** [comment]
Reopening so we have something to backport, what is mentioned above won't fit for backports.

**Comment by andrerom** [comment]
@bdunogier / @pspanja ready for review.

@bdunogier Do you have some pointers on how bdd testing can be done here? aka create -> check -> trash -> check -> recovery -> check

File: eZ/Publish/Core/MVC/Symfony/Cache/GatewayCachePurger.php [hunk 2]
```diff
@@ -21,19 +21,20 @@
...
      */
-    public function purge($cacheElements);
+    public function purge($locationIds);
 
     /**
      * Triggers cache purge for given content.
      * If given content has several locations, cache will be purged for all of them.
      *
      * @param mixed $contentId Content ID.
+     * @param array $locationIds
```
Comment (bdunogier) [inline eZ/Publish/Core/MVC/Symfony/Cache/GatewayCachePurger.php]: Maybe update the comment above, as it refers to this parameter.

**Comment by bdunogier** [comment]
I can see at least three `@todo` in the code, are they safe/intended for merge ?

Also, remember that this is against 6.1.

About behat, the first thing to do would be to phrase the scenario / describe the expected behavior. Note as well that we do have that much behat for REST. Most of the coverage is done with the REST functional tests (behat + HTTP client).

**Comment by andrerom** [comment]
> I can see at least three @todo in the code, are they safe/intended for merge ?

At least two is there to clarify intent of where this is going later (feature wise, not for backports), the one on SingleSlot service should be dropped as it will be able to clear relevant cache tagged with content id once we have multi tag.

> Also, remember that this is against 6.1.

yeah, wanted to reuse PR and thought it might not matter if a patch hit 6.1 even if not tag is planned. You prefer new PR against 6.4 or 6.3?

> About behat, the first thing to do would be to phrase the scenario / describe the expected behavior. 

Maybe @ezsystems/qa-team can help with that? QA; the issues this is potentially solving is:
- https://jira.ez.no/browse/EZP-25003
- https://jira.ez.no/browse/EZP-24621
- https://jira.ez.no/browse/EZP-24944
- https://jira.ez.no/browse/EZP-26107

> Note as well that we do not have that much behat for REST. Most of the coverage is done with the REST functional tests (behat + HTTP client).

I know, but you distinctively had behat as a todo in the PR here :)

**Comment by andrerom** [comment]
Moved to https://github.com/ezsystems/ezpublish-kernel/pull/1742

**Pull request closed without merging**